libcudacxx/docs/extended_api/memory_resource.md

```diff
@@ -0,0 +1,220 @@
+## memory_resource
+
+The `cuda::memory_resource` header provides two main features:
+* The `cuda::resource` and `cuda::resource_with` concepts that provide proper constraints for arbitrary memory resoures
+* The `cuda::resource_ref` resource wrapper as a type erased memory resource, that allows for simple efficient interfaces
+
+These features are an evolution of `std::pmr::memory_resource` that was introduced in C++17. While `std::pmr::memory_resource` provides a polymorphic memory resource that can be adopted through inheritance, it is not properly suited for heterogeneous systems. With the current design it ranges from cumbersome to impossible to verify whether a memory resource provides allocations that are e.g. accessible on device, or whether it can utilize other allocation mechanisms.
+
+### Properties
+
+With C++ we want the type system to verify statically known properties of our code. If a memory resource provides allocations that are device accessible, we do not want to rely on runtime asserts to check whether we can indeed call a device function with that memory. Therefore, we are introducing a property system that is similar to rust traits.
```

Comment:
I don't know if I'd go this far with the comparison to traits, but if we do want to claim that, capitalize "Rust".

Reply:
Removed that comparison